### PR TITLE
[EVM] Support decoding of all previous block types

### PR DESCRIPTION
**Pull request by sideninja**
Support decoding of all previous block types. Block types were changing during development and since it was deployed on previewnet we must make sure we can decode those previously stored blocks.

### CODE REVIEW

File: fvm/evm/types/block.go [hunk 1]
```diff
@@ -141,3 +123,145 @@ var GenesisBlock = &Block{
 }
 
 var GenesisBlockHash, _ = GenesisBlock.Hash()
+
+// Below block type section, defines earlier block types,
+// this is being used to decode blocks that were stored
+// before block type changes. It allows us to still decode
+// a block that would otherwise be invalid if decoded into
+// latest version of the above Block type.
```
Comment (peterargue) [inline fvm/evm/types/block.go:131]: Do we need to keep these forever or just until previewnet is shutdown?

Can you include a note about that in this comment.

File: fvm/evm/types/block.go [hunk 2]
```diff
@@ -141,3 +123,145 @@ var GenesisBlock = &Block{
...
+	ReceiptRoot       gethCommon.Hash
+	TransactionHashes []gethCommon.Hash
+}
+
+// decodeBlockBreakingChanges will try to decode the bytes into all
+// previous versions of block type, if it succeeds it will return the
+// migrated block, otherwise it will return nil.
+func decodeBlockBreakingChanges(encoded []byte) *Block {
+	b0 := &blockV0{}
+	if e := gethRLP.DecodeBytes(encoded, b0); e == nil {
```
Comment (peterargue) [inline fvm/evm/types/block.go:201]: use `err` for clarity. here and elsewhere
```suggestion
	if err := gethRLP.DecodeBytes(encoded, b0); err == nil {
```

**Review by peterargue — APPROVED** [review]

**Review by ramtinms — APPROVED** [review]
looks good, maybe just add a todo to remove it later

**Review by janezpodhostnik — APPROVED** [review]

**Comment by sideninja** [comment]
> looks good, maybe just add a todo to remove it later

Yes, but depends if we can. Not sure what is the status on testnet.

**Pull request merged**